src/docfx/build/toc/ServicePageGenerator.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    /// <summary>
+    /// Generate Service Page.
+    /// </summary>
+    internal class ServicePageGenerator
+    {
+        private readonly PathString _docsetPath;
+        private readonly Input _input;
+        private readonly JoinTOCConfig _joinTOCConfig;
+        private readonly Output _output;
+
+        public ServicePageGenerator(
+            PathString docsetPath,
+            Input input,
+            JoinTOCConfig joinTOCConfig,
+            Output output)
+        {
+            _docsetPath = docsetPath;
+            _input = input;
+            _joinTOCConfig = joinTOCConfig;
+            _output = output;
+        }
+
+        public void GenerateServicePageFromTopLevelTOC(
+            TableOfContentsNode node,
+            ConcurrentBag<FilePath> results,
+            string directoryName = "")
+        {
+            if (node == null || string.IsNullOrEmpty(node.Name))
```

Comment:
nit: node can never be null
```suggestion
            if (string.IsNullOrEmpty(node.Name))
```

Reply:
Fixed.